### PR TITLE
reside-234 Add optional transform to control values

### PR DESCRIPTION
**Pull request by EmmaLRussell**
Allows control metadata to optionally specify a jsonata transform to apply to emitted values.

### CODE REVIEW

**Review by mwoodbri — COMMENTED** [review]
Absolutely no issue with the implementation here, but not entirely convinced that applying transformations should be within scope of form. Feels a bit asymmetrical i.e. pass in 100 get 1 back? Not sure how that fits with `v-model` etc. Could shifting responsibility to parent app to transform values before it uses them be preferable? But certainly defer to your judgement, as you know exactly how this will fit into comet.

**Comment by EmmaLRussell** [comment]
> Absolutely no issue with the implementation here, but not entirely convinced that applying transformations should be within scope of form. Feels a bit asymmetrical i.e. pass in 100 get 1 back? Not sure how that fits with `v-model` etc. Could shifting responsibility to parent app to transform values before it uses them be preferable? But certainly defer to your judgement, as you know exactly how this will fit into comet.

It is asymmetrical, and it abandons the link between the values embedded in the metadata and the values actually emitted. 

It doesn't impact v-model, because that corresponds to the full metadata, with embedded control values, not the emitted serialised values which this affects. So the controls (which don't deal with the transforms) continue to use and update the untransformed embedded values through v-model. 

I agree that it would be better on principle to keep the transforms out of the form, but I think the alternative would be kind of messy - it would mean the app would need to parse the form metadata to discover the transforms to apply, or keep the transforms in a separate part of the metadata. Apart from the shallow dive to apply 'collapsible' we've managed to avoid the app knowing anything about form metadata structure (and I'm hoping to keep it that way, by using an initial form submit to prompt emit of initial parameter values too).

I also think it's likely that percentage-type transforms may turn out to be useful in other applications. So I'm tending to this approach, though I share your concerns!

**Comment by mwoodbri** [comment]
> > Absolutely no issue with the implementation here, but not entirely convinced that applying transformations should be within scope of form. Feels a bit asymmetrical i.e. pass in 100 get 1 back? Not sure how that fits with `v-model` etc. Could shifting responsibility to parent app to transform values before it uses them be preferable? But certainly defer to your judgement, as you know exactly how this will fit into comet.
> 
> It is asymmetrical, and it abandons the link between the values embedded in the metadata and the values actually emitted.
> 
> It doesn't impact v-model, because that corresponds to the full metadata, with embedded control values, not the emitted serialised values which this affects. So the controls (which don't deal with the transforms) continue to use and update the untransformed embedded values through v-model.
> 
> I agree that it would be better on principle to keep the transforms out of the form, but I think the alternative would be kind of messy - it would mean the app would need to parse the form metadata to discover the transforms to apply, or keep the transforms in a separate part of the metadata. Apart from the shallow dive to apply 'collapsible' we've managed to avoid the app knowing anything about form metadata structure (and I'm hoping to keep it that way, by using an initial form submit to prompt emit of initial parameter values too).
> 
> I also think it's likely that percentage-type transforms may turn out to be useful in other applications. So I'm tending to this approach, though I share your concerns!

Thanks... you've convinced me!

**Review by mwoodbri — APPROVED** [review]

File: src/DynamicForm.vue [hunk 1]
```diff
@@ -34,6 +35,7 @@
 
     interface Methods {
         buildValue: (control: DynamicControl) => string | string[] | number | null
+        transformValue(value: string | string[] | number | null, transform: string): any
```
Comment (Nicolas-Dolan) [inline src/DynamicForm.vue]: Perhaps the built value could be given its own custom type to avoid repetition. Not super necessary though.

**Review by Nicolas-Dolan — APPROVED** [review]
Looks good! It makes me wonder whether something similar could be done with the table values/data presentation in general in hint. I have one minor suggestion below.

File: src/DynamicForm.vue (hunk 1 above)
Comment (EmmaLRussell) [reply src/DynamicForm.vue]: Good idea, will add that

**Pull request merged**